Load SSL certificate according to git config

turbonaitis opened the pull request:
This PR addresses #487.

It covers the scenarios, when sslCert configured in .gitconfig points either to a certificate with private key in a keychain, or to a file, containing both public and private keys.

One thing missing is sslKey support - if our .gitconfig specifies sslCert, containing just the public part, and sslKey, containing the private key, the current PR doesn't load it. Main reason being, is that dotnet currently does not support easy handling of PEM files. I'll create a separate PR addressing that, as I believe it will lead to much more discussins.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/VFSForGit?pullRequest=512) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/VFSForGit?pullRequest=512) before we can accept your contribution.<br/><br/>:x: turbonaitis [sign now](https://cla.opensource.microsoft.com/Microsoft/VFSForGit?pullRequest=512)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/VFSForGit?pullRequest=512) it.</sub>

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/VFSForGit?pullRequest=512) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/VFSForGit?pullRequest=512) before we can accept your contribution.<br/><br/>:x: turbonaitis [sign now](https://cla.opensource.microsoft.com/Microsoft/VFSForGit?pullRequest=512)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/VFSForGit?pullRequest=512) it.</sub>

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/VFSForGit?pullRequest=512) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/VFSForGit?pullRequest=512) before we can accept your contribution.<br/><br/>:x: turbonaitis [sign now](https://cla.opensource.microsoft.com/Microsoft/VFSForGit?pullRequest=512)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/VFSForGit?pullRequest=512) it.</sub>

jrbriggs commented on GVFS/GVFS.Common/Http/HttpRequestor.cs:
```diff
@@ -31,22 +41,51 @@ static HttpRequestor()
             this.Tracer = tracer;
 
+            var httpClientHandler = new HttpClientHandler() { UseDefaultCredentials = true };
+            httpClientHandler.ClientCertificateOptions = ClientCertificateOption.Manual;
```
I don't think we should change anything about the mainline flow. Can we only set this when there is a cert provided?

turbonaitis replied:
Moved the ClienCertificateOption assignment further down, to happen just before setting the certificate. The httpClientHandler isn't really new - it was just being created inline, in the HttpClient constructor.

jrbriggs commented on GVFS/GVFS.Common/Http/HttpRequestor.cs:
```diff
@@ -31,22 +41,51 @@ static HttpRequestor()
         }
 
-        public HttpRequestor(ITracer tracer, RetryConfig retryConfig, GitAuthentication authentication)
+        protected HttpRequestor(ITracer tracer, RetryConfig retryConfig, Enlistment enlistment)
```
Should the ssl settings be a property of the authentication object?

turbonaitis replied:
Moved it there.

jrbriggs commented on GVFS/GVFS.Common/Git/GitSslSettings.cs:
```diff
@@ -0,0 +1,40 @@
+            else
+            {
+                this.SslCertificate = default(string);
+                this.SslCertPasswordProtected = default(bool);
```
Let's be explicit about `null` and `false` as the default values.

turbonaitis replied:
Great suggestion, changes made.

jrbriggs commented on GVFS/GVFS.Common/Git/GitProcess.cs:
```diff
@@ -158,6 +158,58 @@ public virtual void RevokeCredential(string repoUrl)
+
+            using (ITracer activity = tracer.StartActivity("TryGetCertificatePassword", EventLevel.Informational))
+            {
+                Result gitCredentialOutput = this.InvokeGitAgainstDotGitFolder(
```
Have you tested this on Windows?

turbonaitis replied:
No, sadly I do not have access to a windows machine. Do you think something in particular may behave differently here?

jrbriggs replied:
Prompting for a password from potentially a background process is just a flow we don't have good automation around, which is why I was asking.

turbonaitis replied:
IMO, this doesn't differ from `TryGetCredentials`, a method you already have. It invokes the same git credential helper, with the same potential of being invoked from a background process. So the way I see it, I'm not introducing anything new here. And yes, when credentials change, while VFS4G is running, and when getting new credentials requires user interaction in terminal (be it for certificate or for http), bad things will happen, but that's something, that is already in the code.

turbonaitis replied:
@wilbaker do you agree with this? Or do you think that my changes introduce some additional risks?

kewillford commented on GVFS/GVFS.Common/Http/HttpRequestor.cs:
```diff
@@ -249,10 +303,42 @@ private static string GetSingleHeaderOrEmpty(HttpHeaders headers, string headerN
+        {
+            if (File.Exists(certId))
+            {
+                return new X509Certificate2(certId, certificatePassword);
```
Do we need some exception handling in this method since this can throw an exception?

turbonaitis replied:
Good point, added.

turbonaitis commented on GVFS/GVFS.Common/Git/GitProcess.cs:
```diff
@@ -253,6 +305,19 @@ public Result SetInFileConfig(string configFile, string settingName, string valu
+        public bool TryGetConfigUrlMatch(string section, string repositoryUrl, out Dictionary<string, GitConfigSetting> configSettings)
+        {
+            Result result = this.InvokeGitAgainstDotGitFolder($"config --get-urlmatch {section} {repositoryUrl}");
+            if (result.HasErrors)
```
Ideally, we would, but I was following the pattern, established by other TryGetConfig methods, which don't take. Also, neither Enlistment constructor (current call site), nor GitAuthentication constructor (potential new call site) have ITracer instance to pass here.

turbonaitis commented on GVFS/GVFS.Common/Http/HttpRequestor.cs:
```diff
@@ -21,8 +23,16 @@ public abstract class HttpRequestor : IDisposable
+
+        private readonly Lazy<X509Store> store = new Lazy<X509Store>(() =>
+        {
+            var s = new X509Store();
```
Out of interest, is there a reason for this? Also, do you guys have some coding guidelines, which I could follow in the future?

kewillford replied:
>is there a reason for this?

We mainly do this for code readability when doing code reviews so you don't have to go searching for the type that is declared.  For example if you have the following:
```
var a = SomeMethod();
<a bunch of code>
var b = a;
<a bunch of code>
var c = b;
```
Doing a code review, you would have to go searching to know what the type of a, b, or c is.

There are also subtle bugs that can slip through.  

I know that in most cases the compiler will catch things and in your case here the type is right there for the reviewer to see, but to be more explicit we have opted to not use it.
>do you guys have some coding guidelines, which I could follow in the future?

We have mostly relied on stylecop to catch any style type issues but that is a good idea of something to add.  I added #537 to track this.

turbonaitis commented on GVFS/GVFS.Common/Git/GitConfigHelper.cs:
```diff
@@ -79,13 +79,14 @@ public static bool TrySanitizeConfigFileLine(string fileLine, out string sanitiz
             Dictionary<string, GitConfigSetting> configSettings = new Dictionary<string, GitConfigSetting>(StringComparer.OrdinalIgnoreCase);
             foreach (string line in input)
             {
-                string[] fields = line.Trim().Split(new[] { '=' }, 2, StringSplitOptions.None);
```
The reason I removed trim is to cater for cases, where the separator is a space and the value is empty. We are trimming `key` and `value` below.

turbonaitis commented on GVFS/GVFS.Common/Enlistment.cs:
```diff
@@ -46,9 +46,18 @@ public abstract class Enlistment
+
+            this.Authentication = authentication ?? new GitAuthentication(gitProcess, this.RepoUrl);
+
+            if (gitProcess.TryGetConfigUrlMatch("http", this.RepoUrl, out var configSettings))
```
Moved `SslSettings` property and its assignment to `GitAuthentication` class.

turbonaitis commented on GVFS/GVFS.Common/Git/GitProcess.cs:
```diff
@@ -158,6 +158,58 @@ public virtual void RevokeCredential(string repoUrl)
         }
 
+        /// <summary>
+        /// Input for certificate crednetials looks like
```
Fixed.

turbonaitis commented on GVFS/GVFS.Common/Git/GitProcess.cs:
```diff
@@ -158,6 +158,58 @@ public virtual void RevokeCredential(string repoUrl)
+
+                if (gitCredentialOutput.HasErrors)
+                {
+                    EventMetadata errorData = new EventMetadata();
```
Added.

sanoursa commented on GVFS/GVFS.Common/Http/HttpRequestor.cs:
```diff
@@ -31,22 +42,51 @@ static HttpRequestor()
             this.Tracer = tracer;
 
+            var httpClientHandler = new HttpClientHandler() { UseDefaultCredentials = true };
+#if DEBUG
```
I don't think the other reviews have called this out yet. Please avoid precompiler constants - you'll notice that we don't have a single one in any of our C# files.

Two other techniques that you can use: 
* git config setting that enables/disables self-signed certs
* Call `ProcessHelper.IsDevelopmentVersion`. This is dangerous though, because we don't want to ship different code than what we test, so I would personally favor a bool config setting that just says yes/no to the question of self-signed certs

I would also encapsulate all of the logic related to self-signed certs into one place. Right now I see a sprinkling of these `#if DEBUG` checks, and that is going to be difficult to maintain.

turbonaitis replied:
Git actually already has http.sslVerify setting, which I'll reuse for this.

turbonaitis commented:
@jrbriggs I think I've addressed all of the comments above. Are there any other outstanding actions, that I should take, before we can merge this PR?

turbonaitis commented on GVFS/GVFS.Common/Http/HttpRequestor.cs:
```diff
@@ -249,10 +306,63 @@ private static string GetSingleHeaderOrEmpty(HttpHeaders headers, string headerN
+                return password;
+            }
+
+            return null;
```
We will try to load the certificate without the password. If that fails, we'll return null for certificate, which means, that we'll try to access our HTTP endpoint without it, which will likely result in an authentication error. 
If there was an error, while retrieving password, it will be logged. 

Another way would be to stop the flow immediately, if either password or certificate can not be retrieved, but I think allowing the customer to proceed till he actually fails a request due to the missing certificate is less disruptive. 

One thing we should probably do, is to log a warning, stating that password was configured, but not retrieved (and same for certificate).

turbonaitis replied:
Added the warning messages

jrbriggs commented:
There's a couple of stylecop warnings to fix.

kewillford commented on GVFS/GVFS.Common/Http/HttpRequestor.cs:
```diff
@@ -21,8 +24,16 @@ public abstract class HttpRequestor : IDisposable
+            X509Store s = new X509Store();
+            s.Open(OpenFlags.OpenExistingOnly | OpenFlags.ReadOnly);
+            return s;
+        });
```
Can we move most of this code closer to where it belongs?  For example this `store` should be in the `GitSslSettings` class along with the `LoadCertificatePassword` and `LoadCertificate`.  Probably want to rename the class to `GitSsl` since it would be more than just the settings.

Then I would have a method on the `GitAuthentication` class that is 
```
void SetupSslIfNeeded(ITracer tracer, HttpClientHandler httpClientHandler, GitProcess gitProcess)
```
Which does all the work in the constructor of this class so it would only have 
```
HttpClientHandler httpClientHandler = new HttpClientHandler() { UseDefaultCredentials = true };
enlistment.Authentication.SetupSslIfNeeded(tracer, httpClientHandler, enlistment.CreateGitProcess());
```

turbonaitis replied:
Great suggestions, will do!

kewillford commented on GVFS/GVFS.Common/Http/HttpRequestor.cs:
```diff
@@ -31,22 +42,63 @@ static HttpRequestor()
 
             this.Tracer = tracer;
 
+            var httpClientHandler = new HttpClientHandler() { UseDefaultCredentials = true };
```
Please remove all uses of `var` since we don't use that in this code base.  Thanks!

turbonaitis replied:
Sure, went and rooted out every one of them in the files, that I've touched.

kewillford commented on GVFS/GVFS.Common/Http/HttpRequestor.cs:
```diff
@@ -249,10 +316,69 @@ private static string GetSingleHeaderOrEmpty(HttpHeaders headers, string headerN
             return string.Empty;
         }
+
+        private string LoadCertificatePassword(string certId, GitProcess git)
```
This should be `GetCertificatePassword` I think because we aren't really loading it anywhere.  Same with `LoadCertificate`.

turbonaitis replied:
Renamed to `GetCertificactePassword`, but I'm not sure I agree with suggestion to rename `LoadCertificate`, as we are loading it - from file or from certificate store. Of course, this is your code base, so if you feel strongly about this, I can rename it too.

kewillford replied:
I prefer to use `Get` because we are getting it either from a file or from the certificate store.  The difference to me is `Get` implies a return value.  `Load` to me doesn't necessarily imply a return value. It could be loading it into a class variable or something else.

turbonaitis replied:
Makes sense, renamed LoadCertificate to GetCertificate.

kewillford commented on GVFS/GVFS.Common/Http/HttpRequestor.cs:
```diff
@@ -249,10 +316,69 @@ private static string GetSingleHeaderOrEmpty(HttpHeaders headers, string headerN
+                { "shouldVerify", onlyLoadValidCertificateFromStore }
+            };
+
+            if (File.Exists(certId))
```
What is the purpose of this check?  Null or empty was already checked before this method was called and the file could be deleted between this call and the constructor for `X509Certificate2` and you are already handling the exception for that but it will be a different code path.  

If the file doesn't exist you run `Find`. If it was deleted between this call and the `X509Certificate2` constructor which means it doesn't exist it will return null.  Not sure how much this matters, just pointing out the difference in code paths.

kewillford replied:
There might be a way to parse the string and determine if the format is a path or name but that seems overly complex so this check will be fine.

turbonaitis replied:
It's too complex, and you can't determine, if a string "certificate" refers to a file in your current working directory, or to a subject name in a certificate store.

kewillford commented on GVFS/GVFS.Common/Http/HttpRequestor.cs:
```diff
@@ -249,10 +316,69 @@ private static string GetSingleHeaderOrEmpty(HttpHeaders headers, string headerN
+            try
+            {
+                var findResults = this.store.Value.Certificates.Find(X509FindType.FindBySubjectName, certId, onlyLoadValidCertificateFromStore);
+                if (findResults?.Count > 0)
```
If there is more than 1, do we care?  How do we know we are picking the right one?

kewillford replied:
There is another option of:
3. Try each one until one works or fail if none work, so add a foreach loop.

Reading up on this and don't we want to use [`FindBySubjectDistinguishedName`](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509findtype?view=netframework-4.7.2)?  The documentation says that `FindBySubjectName`
>will find all certificates with the subject name containing that string, regardless of other subject values. Searching by distinguished name is a more precise search.

To me that mean that if I pass "test", I will get the certificates for "mytest", "test", "testing", etc.  

So option 3 would be the only one to work in that case.  Do we have the full name of the cert so we can use `FindBySubjectDistinguishedName`?

kewillford replied:
I like your plan of logging a warning and narrowing down the certs.  Like you said hopefully there will be enough information in the logs to find out what is going on.  Thanks

turbonaitis replied:
Pushed changes, implementing the above.

kewillford commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,119 @@
+    {
+        public readonly string SslCertificate;
+        public readonly bool SslCertPasswordProtected;
+        public readonly bool SslVerify;
```
Some naming suggestions:  Drop `Ssl`
`CertificatePathOrName`
`IsCertPasswordProtected`
`ShouldVerify`

turbonaitis commented:
Rebased on latest master to prevent merge-conflicts.

kewillford commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,171 @@
+        {
+            EventMetadata metadata = new EventMetadata
+            {
+                { "CertificatePathOrSubjectCommonName", this.CertificatePathOrSubjectCommonName },
```
Use `nameof(this.CertificatePathOrSubjectCommonName)` to avoid the string getting out of sync with the property name.

turbonaitis replied:
It depends - do we want to keep the logs stable, regardless of potential renames of variables? If so, having a hardcoded string (which diverge from the variable name in the future) would give us that.

kewillford replied:
Yes that is true but if it gets renamed to something extremely different like `CertificateExpirationDate` the name would not match the data so there is a balance there.  I try to use `nameof` by default if the name would just be the name of the property then if the property name ever changes you can decide if it better to keep the old name or allow it to change.  The code does use both hard coded strings and `nameof` so there aren't any hard and fast rules here so I'll leave it up to you.

kewillford commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,171 @@
+
+            var result =
+                this.GetCertificateFromFile(tracer, metadata, certificatePassword, onlyLoadValidCertificateFromStore) ??
+                this.GetCertificateFromStore(tracer, metadata, onlyLoadValidCertificateFromStore);
```
We should be able to remove the `onlyLoadValidCertificateFromStore` variable and use the `ShouldVerify` property right?

kewillford commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,171 @@
+                X509Certificate2Collection findResults = store.Certificates.Find(X509FindType.FindBySubjectName, this.CertificatePathOrSubjectCommonName, onlyLoadValidCertificateFromStore);
+                if (findResults?.Count > 0)
+                {
+                    LogCertificateCounts(tracer, metadata, findResults.OfType<X509Certificate2>(), "Found {0} certificates by provided name. Matching DNs: {1}");
```
I don't think the string with format should be so far away from where it is used in case there is something that needs to change and one needs to pass more that 2 items for formatting.  Can we create the string here instead?
```
IEnumerable<X509Certificate2> certificates = findResults.OfType<X509Certificate2>();
string matchingNames = string.Join(Environment.NewLine, certificates.Select(x => x.Subject));
string logMessage = $"Found {certificates.Count()} certificates by provided name. Matching DNs: {matchingNames}";
```

kewillford commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,171 @@
+
+                    X509Certificate2[] certsWithMatchingCns = findResults
+                        .OfType<X509Certificate2>()
+                        .Where(x => x.HasPrivateKey && Regex.IsMatch(x.Subject, string.Format("(^|,\\s?)CN={0}(,|$)", this.CertificatePathOrSubjectCommonName))) // We only want certificates, that have private keys, as we need them. We also want a complete CN match
```
Not a fan of regex.  They are hard to understand and maintain.  Can we just use
`x.Subject.IndexOf($"CN={this.CertificatePathOrSubjectCommonName}") >= 0` ?

turbonaitis replied:
They are hard to understand, if you don't know them :)  I agree though, that they're not really readable, but they allow us to achieve precise string matching.
Your suggested code would also match "CN={subjectName}AndSomethingElse" as well as "DCN={subjectName}" - so not an exact match.

As always, this is your code and your standards, so if you're fine with imprecise matching, I can implement your suggested changes.

kewillford replied:
We can keep it.  It is pretty simple and should not be a performance concern in this code path.

kewillford commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,171 @@
+                        .OrderByDescending(x => x.Verify()) // Ordering by validity in a descending order will bring valid certificates to the beginning
+                        .ThenBy(x => x.NotBefore) // We take the one, that was issued earliest, first
+                        .ThenByDescending(x => x.NotAfter) // We then take the one, that is valid for the longest period
+                        .ToArray();
```
You can remove the call to `ToArray` and eliminate enumerating the collection multiple times.

turbonaitis replied:
But `ToArray` is giving us precisely that - preventing from enumerating the whole collection multiple times. If we remove the `ToArray` call, we'll evaluate the whole expression three times:
1. When getting the number of results for logging
2. When selecting the subject names for logging
3. On the `return` statement`, when selecting the `FirstOrDefault`

When we do `ToArray`, we only operate on the resulting array in all the three cases, without any ordering, or filtering.

kewillford replied:
Thanks I see that now.

kewillford commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,171 @@
+            X509Store store = null;
+            try
+            {
+                store = new X509Store();
```
Use a using for the store?
```
using (X509Store store = new X509Store())
{
```
and you can remove the `finally` at the end

kewillford commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,171 @@
+                        .ThenByDescending(x => x.NotAfter) // We then take the one, that is valid for the longest period
+                        .ToArray();
+
+                    LogCertificateCounts(tracer, metadata, certsWithMatchingCns, "Found {0} certificates with a private key and an exact CN match. DNs (sorted by priority, will take first): {1}");
```
Same as the first comment for `LogCertificateCounts`.

kewillford commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,171 @@
+            }
+            catch (CryptographicException cryptEx)
+            {
+                metadata.Add("Exception", cryptEx);
```
Call `ToString()` on `cryptEx`.

kewillford commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,208 @@
+
+            loggingFunction(
+                metadata,
+                logMessage);
```
We can get rid of the `loggingFunction` now and just call the `tracer.Related` method in each of the case statements.

kewillford commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,208 @@
+
+        public bool ShouldVerify { get; }
+
+        public string GetCertificatePassword(ITracer tracer, GitProcess git)
```
Looks like this can be private now?

kewillford commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,208 @@
+                        new EventMetadata
+                        {
+                                { "SslCertificate", this.certificatePathOrSubjectCommonName }
+                        },
```
Can this use the `metadata` parameter that is passed in?

kewillford commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,208 @@
+                            string.Format(
+                                "Found {0} certificates by provided name. Matching DNs: {1}",
+                                findResults.Count,
+                                GetSubjectNameLineForLogging(findResults.OfType<X509Certificate2>())));
```
Will you break this single line up and use some temp variables for better readability?  Like
```
string subjects = GetSubjectNameLineForLogging(findResults.OfType<X509Certificate2>());
string logMessage = $"Found {findResults.Count} certificates by provided name. Matching DNs: {subjects}";
LogWithAppropriateLevel(tracer, metadata, findResults.OfType<X509Certificate2>(), logMessage);
```

turbonaitis replied:
Unless you insist - I'd prefer not to. Adding these temp variables will just pollute the code IMO.

kewillford replied:
@wilbaker What are you thoughts?

wilbaker replied:
I'm also not a fan of the multi-line indent, but I would recommend the following:

- Push the formatting work down to the helper method
- Put all of the interesting data into an `EventMetadata`
- Remove `GetSubjectNameLineForLogging `

*Example calling code*

```
LogWithAppropriateLevel(
    tracer, 
    metadata, 
    findResults.OfType<X509Certificate2>(), 
    "Found certificates by provided name");
```

And then move the certificate related formatting into `LogWithAppropriateLevel`

```
private static void LogWithAppropriateLevel(ITracer tracer, EventMetadata metadata, IEnumerable<X509Certificate2> certificates, string logMessage)
{
    int numberOfCertificates = certificates.Count();

    EventMetadata certificateMetadata = new EventMetadata(metadata);
    certificateMetadata.Add("CertificateCount", numberOfCertificates);
    certificateMetadata.Add("CertificateDNs", string.Join(Environment.NewLine, certificates.Select(x => x.Subject)));

    switch (numberOfCertificates)
    {
        case 0:
            tracer.RelatedError(certificateMetadata, logMessage);
            break;
        case 1:
            tracer.RelatedInfo(certificateMetadata, logMessage);
            break;
        default:
            tracer.RelatedWarning(certificateMetadata, logMessage);
            break;
    }
}
```

@kewillford what do you think?

wilbaker commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,208 @@
+            {
+                if (configSettings.TryGetValue(GitConfigSetting.HttpSslCert, out GitConfigSetting sslCerts))
+                {
+                    this.certificatePathOrSubjectCommonName = sslCerts.Values.Single();
```
`Single` can throw.  Do we have a guarantee at this point that there is exactly one item in `sslCerts.Values`?  If not, do we need to handle the exception (or check that there's exactly one item first)?

(Same comment applies elsewhere in this method where we're using `Single`)

turbonaitis replied:
Currently, if you have multiple git settings, like
```
http.sslcert = cert1
http.sslcert = cert2
```

Git uses the last one. With that in mind, I'll replace Single() with Last(), although with current git behavior, it should only output one line per config key anyway.

wilbaker commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,208 @@
+
+                if (configSettings.TryGetValue(GitConfigSetting.HttpSslCertPasswordProtected, out GitConfigSetting isSslCertPasswordProtected))
+                {
+                    this.isCertificatePasswordProtected = isSslCertPasswordProtected.Values.Select(bool.Parse).Single();
```
Similar question for `bool.Parse`, what will happen if it throws?

turbonaitis replied:
If we have this git config
```
git config --get-urlmatch http https://localhost:8081 

http.sslcert testCert2 test
http.sslverify fake news
```

Then when you invoke git, this is what you get
```
git ls-remote https://localhost:8081
fatal: bad numeric config value 'fake news' for 'http.sslverify': invalid unit
```

So we should fail, but the question is, if we should catch, and log a friendlier message.

wilbaker replied:
> the question is, if we should catch, and log a friendlier message

Correct me if I'm wrong but it looks like:

- `GitSsl` is only created in the `GitAuthentication` constructor
- The `GitAuthentication` constructor is only called from the `Enlistment` constructor.

I've taken a look at some of the paths that hit the `Enlistment` constructor, and it looks like they *only* expect it to throw `InvalidRepoException` (and they expect the exception to contain a helpful error message).

*Example from `Enlistment` constructor*

```
GitProcess.ConfigResult originResult = gitProcess.GetOriginUrl();
if (!originResult.TryParseAsString(out string originUrl, out string error))
{
    throw new InvalidRepoException("Could not get origin url. git error: " + error);
}

if (originUrl == null)
{
    throw new InvalidRepoException("Could not get origin url. remote 'origin' is not configured for this repo.'");
}
```

I think we should maintain that contract (only throwing helpful `InvalidRepoException` messages from the `Enlistment` constructor), and so I think we have a few options:

- Move the code that can throw out of the constructor and into a `TryInitialize` type method
- Update `GitSsl` to catch the exception that can get thrown in its constructor and throw `InvalidRepoException` exceptions instead (with a helpful error message)

We could also catch everything in the `Enlistment` constructor and convert it to a `InvalidRepoException`, but my concern is that message will not have as much context for the user (and these are user facing messages):

*Example of code that handles `InvalidRepoException`*

```
catch (InvalidRepoException e)
{
    this.ReportErrorAndExit(
        "Error: '{0}' is not a valid GVFS enlistment. {1}",
        enlistmentRootPath,
        e.Message);
}
```

turbonaitis replied:
Makes sense, will make sure we throw exceptions with messages, that are helpful to the user.

turbonaitis replied:
Looked at uses - it seems CloneVerb only handles generic exceptions, not `InvalidRepoException`:
https://github.com/Microsoft/VFSForGit/blob/master/GVFS/GVFS/CommandLine/CloneVerb.cs#L244-L261

Added handling for this exception and started throwing it from GitSsl constructor, when bool parsing fails.

wilbaker commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,208 @@
+
+            if (result == null)
+            {
+                tracer.RelatedError("Certificate {0} not found", this.certificatePathOrSubjectCommonName);
```
Pass in `metadata` here?

wilbaker commented on GVFS/GVFS.Common/Git/GitAuthentication.cs:
```diff
@@ -152,6 +166,25 @@ public bool TryInitializeAndRequireAuth(ITracer tracer, out string errorMessage)
             return false;
         }
 
+        public void SetupSslIfNeeded(ITracer tracer, HttpClientHandler httpClientHandler, GitProcess gitProcess)
```
Naming suggestion: 

When I first read `SetupSslIfNeeded` I thought this method was doing something to setup `this.GitSsl`.  Assuming my understanding is correct it actually looks like the method configures ssl on the `HttpClientHandler`.

If that's the case consider renaming this to something like `ConfigureHttpClientHandlerSSLIfNeeded`.  If my understanding is not correct, please point me in the right direction 😄

turbonaitis replied:
Sure, will make the name more explicit. Just FYI, this exact name was suggested by one of the other reviewers, so I hope he doesn't come back asking me to revert it.

wilbaker commented on GVFS/GVFS.UnitTests/Git/GitAuthenticationTests.cs:
```diff
@@ -173,6 +175,7 @@ private MockGitProcess GetGitProcess()
             MockGitProcess gitProcess = new MockGitProcess();
             gitProcess.SetExpectedCommandResult("config gvfs.FunctionalTests.UserName", () => new GitProcess.Result(string.Empty, string.Empty, GitProcess.Result.GenericFailureCode));
             gitProcess.SetExpectedCommandResult("config gvfs.FunctionalTests.Password", () => new GitProcess.Result(string.Empty, string.Empty, GitProcess.Result.GenericFailureCode));
+            gitProcess.SetExpectedCommandResult("config --get-urlmatch http mock://repoUrl", () => new GitProcess.Result($"http.sslCert {CertificatePath}\nhttp.sslCertPasswordProtected true\n\n", string.Empty, GitProcess.Result.SuccessCode));
```
It looks like with this change we won't get any coverage for the scenario where `http.sslCert` is not in the configuration.

Could you parameterize these tests so that we can get coverage for both scenarios? (see `DiffHelperTests` for an example of where we've parameterized unit tests in the past).

wilbaker commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,208 @@
+                metadata.Add("isPasswordSpecified", string.IsNullOrEmpty(certificatePassword));
+            }
+
+            if (File.Exists(this.certificatePathOrSubjectCommonName))
```
For unit testing `PhysicalFileSystem.FileExists` should be used instead of `File.Exists`

wilbaker commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,208 @@
+
+namespace GVFS.Common.Git
+{
+    public class GitSsl
```
Add unit tests for this class?

turbonaitis replied:
I'd love to, but it's a bit challenging, due to the fact, that certificate store has no interface, that we could mock and I don't think we want to write certificates to the actual store during unit testing, as this might not work on CI servers. Also, the certificates are being validated using native libraries (that is - it's outside our control), so it would be very hard to setup a valid certificate. If you have ideas how to approach this - they would be appreciated.

wilbaker replied:
Those are great points.  @kewillford, @jrbriggs, do you have any thoughts on additional tests that could be added to validate `GitSsl`?

My main concern is that we might inadvertently break this feature\functionality if we don't have sufficient coverage of it in our unit and functional tests.

turbonaitis replied:
Absolutely, covering this with tests is something, that I think all of us want. One way to achieve this is to add additional layers of abstraction, like:
```
interface ICertificateLoader
{
   X509Certificate2 LoadCertificateBySubjectName(string subjectName);
}

interface ICertificateVerifier
{
   bool VerifyCertificate(X509Certificate2 certificate);
}
```
Then use them in our GitSsl class and mock those in tests. That's something, that would work, but I'd like to hear your opinion on the solution first, as I didn't see similar code in GVFS code base.

turbonaitis replied:
So, finally got time to get back to this and pushed unit tests for GitSsl. Would you have a look @wilbaker ?

wilbaker commented on GVFS/GVFS.Common/Git/GitProcess.cs:
```diff
@@ -164,6 +164,61 @@ public virtual void RevokeCredential(string repoUrl)
+
+                password = ParseValue(gitCredentialOutput.Output, "password=");
+
+                bool success = password != null;
```
Are empty passwords allowed?

turbonaitis replied:
This is the same behavior as here:
https://github.com/Microsoft/VFSForGit/blob/7390eba79729141c9042b93750a108cdf4dafc2a/GVFS/GVFS.Common/Git/GitProcess.cs#L200

The idea is, that empty password will be an empty string, not a null.

turbonaitis commented on GVFS/GVFS.Common/Git/GitAuthentication.cs:
```diff
@@ -25,6 +28,15 @@ public GitAuthentication(GitProcess git, string repoUrl)
+            }
+            else
+            {
+                this.GitSsl = new GitSsl();
```
It was used more widely, so having default values was good. We can leave it as null and use null propagation now.

wilbaker commented on GVFS/GVFS.Common/GVFS.Common.csproj:
```diff
@@ -28,4 +28,10 @@
+  </ItemGroup>
+  <ItemGroup>
+    <Reference Include="System.Security.Cryptography.X509Certificates" />
+  </ItemGroup>
```
I don't think any of these lines are needed.  I found that with them the build on Windows complains:

```
2019-01-25T18:13:45.7237754Z   C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\amd64\Microsoft.Common.CurrentVersion.targets(2106,5): warning MSB3245: Could not resolve this reference. Could not locate the assembly "System.Security.Cryptography.X509Certificates". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors. [C:\agent\_work\11\s\GVFS\GVFS.Common\GVFS.Common.csproj]
2019-01-25T18:13:45.7238315Z   C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\amd64\Microsoft.Common.CurrentVersion.targets(2106,5): warning MSB3243: No way to resolve conflict between "System.Security.Cryptography.X509Certificates, Version=4.1.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" and "System.Security.Cryptography.X509Certificates". Choosing "System.Security.Cryptography.X509Certificates, Version=4.1.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" arbitrarily. [C:\agent\_work\11\s\GVFS\GVFS.Common\GVFS.Common.csproj]
```

And with them removed the build is able to locate the System.Security.Cryptography.X509Certificates assembly.

wilbaker commented on GVFS/GVFS.Common/X509Certificates/SystemCertificateStore.cs:
```diff
@@ -0,0 +1,28 @@
+
+        public X509Certificate2Collection Find(X509FindType findType, string searchString, bool validOnly)
+        {
+            if (!this.store.IsOpen)
```
This line is causing a compilation failure on Windows (where GVFS.Common builds with .NET Framework):

```
5>------ Build started: Project: GVFS.Common, Configuration: Debug x64 ------
5>X509Certificates\SystemCertificateStore.cs(20,29,20,35): error CS1061: 'X509Store' does not contain a definition for 'IsOpen' and no extension method 'IsOpen' accepting a first argument of type 'X509Store' could be found (are you missing a using directive or an assembly reference?)
5>Done building project "GVFS.Common.csproj" -- FAILED.
```

And based on the context Window I saw in the IDE, it looks like `IsOpen` is not available in the .NET Standard API.

![image](https://user-images.githubusercontent.com/25354985/51765446-a0651480-208c-11e9-8f9c-6f5e6f0fcaa5.png)

Is there a different API you can use to perform this check?

(If you don't have a Windows machine available you can use the required PR builds to check if the build on Windows is successful)

turbonaitis replied:
@wilbaker after I fixed these two, there's another issue, that `GVFS.UnitTests.Windows` uses .NET 4.6.1, which does not have `CertificateRequest` class, which I'm using in my unit tests to generate self signed certificates.  Would it be possible to target 4.7.2 in the unit tests project?

wilbaker replied:
> Would it be possible to target 4.7.2 in the unit tests project?

My preference is to leave the unit tests targeting 4.6.1 as that's what the VFS4G executables on Windows are using (and we're not mocking out everything in the .NET Framework).

Rather than retargeting `GVFS.UnitTests.Windows` what would you think of only running these tests as part of `GVFS.UnitTest`?

The cleanest way I've found for doing that is adding the following to `GitSslTests.cs`:

```
#if NETCOREAPP2_1 // .NET Framework 4.6.1 does not support System.Security.Cryptography.X509Certificates.CertificateRequest

// All of the contents of GitSslTests.cs is here and unchanged

#endif
```

I still need to take a look at the new tests you've added, and will post feedback after taking a look.  Thanks!

CC: @kewillford and @nickgra for their thoughts as well.

nickgra replied:
After some digging through history, I agree that `#if NETCOREAPP2_1` is likely the best solution here even though it's a code smell. 

In the past when we started migrating to .NET Core, we had some tests that broke due to changes in how .NET hashes strings. While the problem was actually an issue with the way the test was written, it concerns me that we could write a UT that would behave differently than the actual product code would if they're targeting different platform versions.

In the medium to long term, I plan to move the entire project (Windows and Mac) to .NET Core 3.0 when it comes out (We need C++/CLI support on Windows, which is going to be supported in 3.0), so I can clean this up when that time comes. I think that's a better investment than us moving forward to 4.7.2 right now.

turbonaitis replied:
Sounds great, will do this!

turbonaitis commented on GVFS/GVFS.UnitTests/Common/Git/GitSslTests.cs:
```diff
@@ -0,0 +1,266 @@
+    {
+        public static object[] BoolGitSettings = new[]
+        {
+            GitConfigSetting.HttpSslCertPasswordProtected,
```
Thanks for the tip, didn't know of this interesting NUnit behavior!

kewillford commented on GVFS/GVFS.UnitTests/Mock/ReusableMemoryStream.cs:
```diff
@@ -16,6 +16,12 @@ public ReusableMemoryStream(string initialContents)
             this.length = this.contents.Length;
         }
 
+        public ReusableMemoryStream(byte[] initialContnets)
```
Spelling

kewillford commented on GVFS/GVFS.Common/Git/GitSsl.cs:
```diff
@@ -0,0 +1,237 @@
+            }
+        }
+
+        private GitSsl(Func<SystemCertificateStore> certificateStoreFactory, CertificateVerifier certificateVerifier, PhysicalFileSystem fileSystem)
```
Parameter is `certificateStoreFactory` and `createCertificateStore` above.  They should probably match.

jrbriggs commented:
Thanks for the contribution, @turbonaitis !
